### PR TITLE
Recompute SDR in float64 type if float32 gives NaN or Inf

### LINKED ISSUE #895: The sdr metric in TM sometime gives NaN for some input
## 🐛 Bug

This issue is related with `fast-bss-eval`'s torch version, see https://github.com/fakufaku/fast_bss_eval/issues/5

### To Reproduce

```
import numpy as np
import torch

x = np.load('debug.npz')
preds = torch.tensor(x['preds'])
target = torch.tensor(x['target'])
print(preds.shape, target.shape)

from torchmetrics.functional.audio import signal_distortion_ratio
sdr = signal_distortion_ratio(preds, target)
print(sdr)

from mir_eval.separation import bss_eval_sources
sdr, _, _, _ = bss_eval_sources(target.numpy(), preds.numpy(), False)
print(sdr)
```

outputs:
```
torch.Size([2, 64000]) torch.Size([2, 64000])
tensor([-2.6815,     nan])
[-2.68156071 44.58523729]
```

unzip [data.zip](https://github.com/PyTorchLightning/metrics/files/8309638/data.zip) to get the `debug.npz`

### Expected behavior

the results given by signal_distortion_ratio is close to the one given by `mir_eval`

### Environment

- OS (e.g., Linux):
- Python & PyTorch Version (e.g., 1.0):
- How you installed PyTorch (`conda`, `pip`, build command if you used source):
- Any other relevant information:

### PR DESCRIPTION
**Pull request by quancs**
## What does this PR do?

Fixes #895

## Before submitting

- [x] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/metrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by quancs** [comment]
> can we add a test for this case...

OK

File: torchmetrics/functional/audio/sdr.py [hunk 1]
```diff
@@ -192,6 +192,11 @@ def signal_distortion_ratio(
     # transform to decibels
     ratio = coh / (1 - coh)
     val = 10.0 * torch.log10(ratio)
+
+    # recompute sdr in float64 if val is NaN or Inf
+    if (torch.isnan(val).any() or torch.isinf(val).any()) and preds.dtype != torch.float64:
+        val = signal_distortion_ratio(preds.double(), target.double(), use_cg_iter, filter_length, zero_mean, load_diag)
```
Comment (SkafteNicki) [inline torchmetrics/functional/audio/sdr.py:198]: lets give the user a warning here
```suggestion
    if (torch.isnan(val).any() or torch.isinf(val).any()) and preds.dtype != torch.float64:
        warnings.warn(
            "Detected `nan` or `inf` value in computed metric, retrying computation in double precision",
            UserWarning,
        )
        val = signal_distortion_ratio(preds.double(), target.double(), use_cg_iter, filter_length, zero_mean, load_diag)
```

File: tests/audio/test_sdr.py [hunk 2]
```diff
@@ -148,3 +149,18 @@ def test_on_real_audio():
         rtol=0.0001,
         atol=1e-4,
     )
+
+
+def test_issue_895():
+    current_file_dir = os.path.dirname(__file__)
```
Comment (SkafteNicki) [inline tests/audio/test_sdr.py:155]:
```suggestion
test_issue_with_too_low_precision():
    """ Corner case where the precision of the input is important. """
    current_file_dir = os.path.dirname(__file__)
```

File: tests/audio/test_sdr.py [hunk 3]
```diff
@@ -148,3 +149,18 @@ def test_on_real_audio():
...
+    data = np.load(os.path.join(current_file_dir, "examples/issue_895.npz"))
+    preds = torch.tensor(data["preds"])
+    target = torch.tensor(data["target"])
+    sdr_tm = signal_distortion_ratio(preds, target)
```
Comment (SkafteNicki) [inline tests/audio/test_sdr.py:159]: if we add the warning we will also be sure that the new code is actually triggered in this test
```suggestion
    with pytest.warns(UserWarning, match="Detected `nan` or `inf` value in computed metric, retrying computation in double precision"):
        sdr_tm = signal_distortion_ratio(preds, target)
```

File: tests/audio/test_sdr.py [hunk 4]
```diff
@@ -148,3 +149,22 @@ def test_on_real_audio():
         rtol=0.0001,
         atol=1e-4,
     )
+
+
+def test_issue_with_too_low_precision():
```
Comment (Borda) [inline tests/audio/test_sdr.py:154]:
```suggestion
def test_too_low_precision():
```

**Review by Borda — APPROVED** [review]

**Review by justusschock — APPROVED** [review]
